components/layout/display_list/largest_contenful_paint_collector.rs

```diff
@@ -0,0 +1,163 @@
+    pub viewport_rect: LayoutRect,
+    /// The current [`Epoch`]
+    pub cur_epoch: Epoch,
+}
+
+impl<'a> LargestContentfulPaintCandidateCollector<'a> {
+    pub fn new(
+        clip_tree: &'a StackingContextTreeClipStore,
+        current_scroll_node_id: ScrollTreeNodeId,
+        is_recording_lcp: bool,
+        viewport_size: LayoutSize,
+        cur_epoch: Epoch,
+    ) -> Self {
+        Self {
+            clip_tree,
+            current_scroll_node_id,
+            current_clip_id: ClipId::INVALID,
+            lcp_candidate: None,
+            is_recording_lcp,
+            viewport_rect: LayoutRect::from_size(viewport_size),
+            cur_epoch,
+        }
+    }
+
+    pub fn update_current_node_id(
+        &mut self,
+        next_scroll_node_id: ScrollTreeNodeId,
+        next_clip_id: ClipId,
+    ) {
+        self.current_scroll_node_id = next_scroll_node_id;
+        self.current_clip_id = next_clip_id;
+    }
+
+    pub fn update_image_candidate(
+        &mut self,
+        id: LCPCandidateID,
+        rect: LayoutRect,
+        scroll_tree: &ScrollTree,
+    ) {
+        if !self.is_recording_lcp {
+            return;
+        }
+
+        let visual_rect = self.compute_visual_rect(rect, scroll_tree);
+        let area = visual_rect.area() as usize;
+        if area == 0 {
+            return;
+        }
+
+        // TODO(boluochoufeng): need to filter low-content image
+
+        self.update_candidate(LCPCandidate::new(id, area, self.cur_epoch));
+    }
+
+    fn update_candidate(&mut self, candidate: LCPCandidate) {
+        if let Some(ref mut latest_candidate) = self.lcp_candidate {
+            if candidate.area > latest_candidate.area {
+                *latest_candidate = candidate;
+            }
+        } else {
+            self.lcp_candidate = Some(candidate);
+        }
+    }
+
+    /// Calculate the size of visual area in viewport. Because the parent elements will affect children,
+    /// We need track these css up to the root node.
+    fn compute_visual_rect(&self, rect: LayoutRect, scroll_tree: &ScrollTree) -> LayoutRect {
+        let visual_rect = self.compute_visual_rect_with_scroll_clip(
+            rect,
+            Some(self.current_scroll_node_id),
+            self.current_clip_id,
+            scroll_tree,
+        );
+
+        visual_rect
+            .intersection(&self.viewport_rect)
+            .unwrap_or(LayoutRect::zero())
+    }
+
+    fn compute_visual_rect_with_scroll_clip(
+        &self,
+        rect: LayoutRect,
+        scroll_id: Option<ScrollTreeNodeId>,
+        clip_id: ClipId,
+        scroll_tree: &ScrollTree,
+    ) -> LayoutRect {
+        let mut visual_rect = rect;
+        let mut parent_clip_id = clip_id;
+        let mut parent_scroll_id = scroll_id;
+
+        while let Some(scroll_id) = parent_scroll_id {
+            // 1.If there is clip CSS in the stacking context created by transform, apply it.
+            if let Some(clip_node) = &self
```

Comment:
Should this be a while loop? There might be multiple clippings between the current node and its nearest scroll container